File: py/sentence_splitter.py
from concord.computation import (
    Computation,
    Metadata,
    serve_computation
)
from concord.internal.thrift.ttypes import StreamGrouping

class SentenceSplitter(Computation):
    def init(self, ctx):
        self.concord_logger.info("Splitter initialized")

    def process_timer(self, ctx, key, time):
        raise Exception('process_timer not implemented')

    def process_record(self, ctx, record):
        for word in record.key.split(" "):
            ctx.produce_record('words', word, '-')

    def metadata(self):
        return Metadata(
            name='sentence-splitter',
            istreams=['sentences'],
            ostreams=['words'])

serve_computation(SentenceSplitter())

File: rb/sentence_splitter.rb
require 'concord'
require 'concord/utils'

class SentenceSplitter
  def init(context)
    Concord::Utils.log_to_stderr("Initialized Splitter computation")
  end

  # Use the destroy method to perform some cleanup before the framework
  # kills this process
  def destroy
    Concord::Utils.log_to_stderr("Destructing splitter computation")
  end

  def process_record(context, record)
    record.key.split do |word|
      context.produce_record('words', word, '-')
    end
  end

  def metadata
    Concord::Metadata.new(name: 'sentence-splitter',
                          istreams: [['sentences', Concord::Thrift::StreamGrouping::SHUFFLE]],
                          ostreams: ['words'])
  end
end

Concord::Computation.serve(SentenceSplitter.new)
